flocker/control/_diffing.py

```diff
@@ -95,6 +100,152 @@ def apply(self, obj):
         return obj.transform(self.path, lambda x: x.add(self.item))
 
 
+_sentinel = object()
+
+
+class _EvolverProxy(object):
+    """
+    This attempts to bunch all the diff operations for a particular object into
+    a single transaction so that related attributes can be ``set`` without
+    triggering an in invariant error.
+    Additionally, the leaf nodes are persisted first and in isolation, so as
+    not to trigger invariant errors in ancestor nodes.
+    """
+    def __init__(self, original):
+        """
+        :param PClass original: The root object to which transformations will
+            be applied.
+        """
+        self._original = original
+        self._evolver = original.evolver()
+        self._children = {}
+        self._operations = []
+
+    def _child(self, segment):
+        child = self._children.get(segment)
+        if child is not None:
+            return child
+        child = _get(self._original, segment, _sentinel)
+        if child is _sentinel:
+            raise KeyError(
+                'Segment not found in path. '
+                'Parent: {}, '
+                'Segment: {}'.format(self, segment)
+            )
+        proxy_for_child = _EvolverProxy(child)
+        self._children[segment] = proxy_for_child
+        return proxy_for_child
+
+    def transform(self, path, operation):
+        """
+        Traverse each segment of ``path`` to create a hierarchy of
+        ``_EvolverProxy`` objects and perform the ``operation`` on the
+        resulting leaf proxy object. This will infact perform the operation on
+        an evolver of the original Pyrsistent object.
+
+        :param PVector path: The path relative to ``original`` which will be
+            operated on.
+        :param callable operation: A function to be applied to an evolver of
+             the object at ``path``
+        :returns: ``self``
+        """
+        target = self
+        for segment in path:
+            target = target._child(segment)
+        operation(target)
+        return self
+
+    def add(self, item):
+        """
+        Add ``item`` to the ``original`` ``Pset`` or if the item is itself a
+        Pyrsistent object, add a new proxy for that item so that further
+        operations can be performed on it without triggering invariant checks
+        until the tree is finally committed.
+
+        :param item: An object to be added to the ``PSet`` wrapped by this
+            proxy.
+        :returns: ``self``
+        """
+        if hasattr(item, 'evolver'):
```

Comment:
`hasattr` can mask errors.  It's safer in general to use `getattr(obj, attr, sentinel)` and compare against sentinel.